src/light_wallet_server/scanner.cpp

```diff
@@ -0,0 +1,756 @@
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#include "scanner.h"
+
+#include <algorithm>
+#include <boost/numeric/conversion/cast.hpp>
+#include <boost/range/combine.hpp>
+#include <boost/thread/condition_variable.hpp>
+#include <boost/thread/mutex.hpp>
+#include <boost/thread/thread.hpp>
+#include <cassert>
+#include <chrono>
+#include <cstring>
+#include <type_traits>
+#include <utility>
+
+#include "common/error.h"
+#include "crypto/crypto.h"
+#include "cryptonote_basic/cryptonote_basic.h"
+#include "cryptonote_basic/cryptonote_format_utils.h"
+#include "cryptonote_core/cryptonote_tx_utils.h"
+#include "device/device_default.hpp"
+#include "light_wallet_server/db/account.h"
+#include "light_wallet_server/db/data.h"
+#include "light_wallet_server/error.h"
+#include "light_wallet_server/json.h"
+#include "misc_log_ex.h"
+#include "rpc/daemon_messages.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "lws"
+
+namespace lws
+{
+    std::atomic<bool> scanner::running{true};
+
+    // Not in `rates.h` - defaulting to JSON output seems odd
+    std::ostream& operator<<(std::ostream& out, lws::rates const& src)
+    {
+        json::rates(out, src);
+        return out;
+    }
+
+    namespace
+    {
+        constexpr const std::chrono::seconds account_poll_interval{10};
+        constexpr const std::chrono::seconds block_poll_interval{20};
+        constexpr const std::chrono::minutes block_rpc_timeout{2};
+        constexpr const std::chrono::seconds send_timeout{30};
+        constexpr const std::chrono::seconds sync_rpc_timeout{30};
+
+        struct thread_sync
+        {
+            boost::mutex sync;
+            boost::condition_variable user_poll;
+            std::atomic<bool> update;
+        };
+
+        struct thread_data
+        {
+            explicit thread_data(rpc::client client, db::storage disk, std::vector<lws::account> users)
+              : client(std::move(client)), disk(std::move(disk)), users(std::move(users))
+            {}
+
+            rpc::client client;
+            db::storage disk;
+            std::vector<lws::account> users;
+        };
+
+        // until we have a signal-handler safe notification system
+        void checked_wait(const std::chrono::nanoseconds wait)
+        {
+            static constexpr const std::chrono::milliseconds interval{500};
+
+            const auto sleep_time = std::min(wait, std::chrono::nanoseconds{interval});
+            const auto start = std::chrono::steady_clock::now();
+            while (scanner::is_running() && (std::chrono::steady_clock::now() - start) < wait)
+                boost::this_thread::sleep_for(boost::chrono::milliseconds{sleep_time.count()});
```

Comment:
I tested this on my machine Ubuntu 18.04 on x64 platform, it actually sleep for 100 000 times longer than wanted.
With the following code to print real sleep time before the while loop:
`
const boost::chrono::milliseconds real_sleep_time = boost::chrono::milliseconds{sleep_time.count()};            MINFO("real_sleep_time(ms): " << real_sleep_time.count());
`
Suggest change it to :
`
std::this_thread::sleep_for(sleep_time);
`

Reply:
I missed this comment somehow. The bug is that this is in nanosecods not milliseconds, so its off by 1000. Will push out a fix soon. Glad to see someone testing/using.